### PR TITLE
`Communication`: Support File Sharing

### PR DESCRIPTION
**Pull request by eylulnc**
### Problem Description
This PR introduces a new feature that allows users to upload files directly to a conversation.

### Changes
 
**File Upload UI:**

Added buttons for file in message area

**Validation:**

Enforced a 5MB file size limit.
Restricted uploads to supported file types.

### Steps for testing
Test the new buttons added to the messsage field for file upload.
Upload a file or image and verify:
Files/images are displayed with appropriate markdown.

### Screenshots

Screenshot 2024-12-14 at 00 32 17

### CODE REVIEW

File: ArtemisKit/Sources/Messages/Services/MessagesService/MessagesServiceImpl.swift [hunk 1]
```diff
@@ -258,6 +258,33 @@ struct MessagesServiceImpl: MessagesService {
         }
     }
 
+    struct UploadFileResult: Codable {
+           let path: String
+       }
+
+       func uploadFile(for courseId: Int, and conversationId: Int64, file: Data, filename: String, mimeType: String) async -> DataState<String> {
```
Comment (anian03) [inline ArtemisKit/Sources/Messages/Services/MessagesService/MessagesServiceImpl.swift:265]: Could we remove `uploadImage` and replace its usages with a call to this function? From what I can see it's the same, just with images using someUUID.jpg as fileName and Image/jpeg as mimeType.

File: ArtemisKit/Sources/Messages/Services/MessagesService/MessagesServiceImpl.swift [hunk 2]
```diff
@@ -258,6 +258,33 @@ struct MessagesServiceImpl: MessagesService {
         }
     }
 
+    struct UploadFileResult: Codable {
+           let path: String
+       }
```
Comment (anian03) [inline ArtemisKit/Sources/Messages/Services/MessagesService/MessagesServiceImpl.swift:263]:
```suggestion
        let path: String
    }
```

File: ArtemisKit/Sources/Messages/ViewModels/SendMessageViewModels/FileUpload/SendMessageUploadFileViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        isFilePickerPresented = false
+        loadFileData(from: url)
+    }
+
+    /// Reads the file data from the provided URL
+    private func loadFileData(from url: URL) {
+        uploadState = .compressing
+        filePath = nil
+
+        Task {
```
Comment (anian03) [inline ArtemisKit/Sources/Messages/ViewModels/SendMessageViewModels/FileUpload/SendMessageUploadFileViewModel.swift:61]: I don't think we need to wrap this block inside a `Task` (we don't use async/await here).

**Review by anian03 — CHANGES_REQUESTED** [review]
Looks pretty good overall now. Just some very small things that would be great if you could improve – otherwise we're good to go 😁

**Comment by anian03** [comment]
Maybe something in general that we could think about for a follow up: Do we want to combine uploading images and files into a single button with a menu (like the + on the left)? I think that's how we do it on the web app, and most other apps do something similar

**Comment by eylulnc** [comment]
> Maybe something in general that we could think about for a follow up: Do we want to combine uploading images and files into a single button with a menu (like the + on the left)? I think that's how we do it on the web app, and most other apps do something similar

I think we can discuss this and depending on the outcome also we create one more ticket for Android as well

**Review by anian03 — APPROVED** [review]
Thanks for the changes!

**Pull request merged**